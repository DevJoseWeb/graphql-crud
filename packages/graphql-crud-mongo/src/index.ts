export * from './MongoStore';
import { MongoStore } from './MongoStore';

export default MongoStore;
