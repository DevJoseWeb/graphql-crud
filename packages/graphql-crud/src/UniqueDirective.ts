import { SchemaDirectiveVisitor } from 'graphql-tools';

export const createUniqueDirective = () => {
  return class UniqueDirective extends SchemaDirectiveVisitor {
  };
};
