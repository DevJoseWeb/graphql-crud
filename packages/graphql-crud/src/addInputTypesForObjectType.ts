import {
  GraphQLInputObjectType,
  GraphQLList,
  GraphQLObjectType,
  GraphQLSchema,
} from 'graphql';
import {
  getInputType,
  isValidInputType,
  omitResolvers,
  toInputObjectTypeName,
} from './';

export interface AddInputTypesForObjectTypeProps {
  objectType: GraphQLObjectType;
  schema: GraphQLSchema;
  prefix?: string;
  modifyField?: (field: any) => any;
}

export const addInputTypesForObjectType = ({
  objectType,
  schema,
  prefix = '',
  modifyField = (field) => field,
}: AddInputTypesForObjectTypeProps) => {
  // Fields of an input type cannot have resolvers
  // console.log(objectType);
  const fields = omitResolvers(objectType.getFields());

  // Create the corresponding input type.
  // For example, if given `type Foo` will create `input FooInputType`
  let inputObjectType = new GraphQLInputObjectType({
    name: `${prefix}${toInputObjectTypeName(objectType.name)}`,
    fields,
  });

  // Adds the newly created input type to the type map.
  //
  // Note: the GraphQLObjectType fields of the input type have not yet been replaced.
  // However weneed a reference to the input type added to the type map for lookups during recursion.
  schema.getTypeMap()[inputObjectType.name] = inputObjectType;

  // Iterate over each field in the input type.
  // If the field's type is not a GraphQLObjectType or a GraphQLList then it is copied as is.
  // If the field's type is a GraphQLObjectType or a GraphQLList
  //  Get the type which the GraphQLList contains
  //  Find (or create if not found) the corresponding input type
  //  Replace the field's type with the input type
  const inputObjectFields = Object
    .keys(fields)
    .reduce((res, key) => {
      let field = fields[key];

      if (!isValidInputType(field.type, schema)) {
        // Check if the input type already exists
        const inputType = getInputType(`${prefix}${field.type.name}`, schema);
        if (inputType) {
          field = {
            name: inputType.name,
            type: field.type instanceof GraphQLList ? new GraphQLList(inputType) : inputType,
          };
        } else {
          // Input type does not exist so we need to create it
          const fieldType = schema.getType(field.type.ofType || field.type.name); // `field.type.ofType` is used in case of a list type
          const newInputType = addInputTypesForObjectType({
            objectType: fieldType as GraphQLObjectType,
            schema,
            prefix,
            modifyField,
         });
          field = {
            name: newInputType.name,
            type: field.type instanceof GraphQLList ? new GraphQLList(newInputType) : newInputType,
          };
        }
      }

      return {
        ...res,
        [key]: modifyField(field),
      };
    }, {});

  // Replace our original inputObjectType with new one containing the modified fields

  inputObjectType = new GraphQLInputObjectType({
    name: inputObjectType.name,
    fields: inputObjectFields,
  });

  schema.getTypeMap()[inputObjectType.name] = inputObjectType;

  return inputObjectType;
};
