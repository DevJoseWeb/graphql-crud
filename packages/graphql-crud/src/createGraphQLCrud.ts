import { createModelDirective, DefaultDirective, RelationDirective, UniqueDirective } from '.';

export const createGraphQLCrud = () => ({
  schemaDirectives: {
    model: createModelDirective(),
    default: DefaultDirective,
    relation: RelationDirective,
    unique: UniqueDirective,
  },
});
