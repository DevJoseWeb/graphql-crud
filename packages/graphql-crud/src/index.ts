export * from './createModelDirective';
export * from './RelationDirective';
export * from './DefaultDirective';
export * from './UniqueDirective';
export * from './Store';
export * from './generateFieldNames';
export * from './omitResolvers';
export * from './addInputTypesForObjectType';
export * from './util';
export * from './validateInputData';
export * from './createGraphQLCrud';
