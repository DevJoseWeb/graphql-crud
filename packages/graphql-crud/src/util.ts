import {
  getNamedType,
  GraphQLInputObjectType,
  GraphQLObjectType,
  GraphQLSchema,
} from 'graphql';

export const toInputObjectTypeName = (name: string): string => `${name}InputType`;

export const isValidInputFieldType = (type): boolean => {
  return !(getNamedType(type) instanceof GraphQLObjectType);
};

export const getInputType = (typeName: string, schema: GraphQLSchema): GraphQLInputObjectType => {
  const type = schema.getType(toInputObjectTypeName(typeName));
  return type as GraphQLInputObjectType;
};
