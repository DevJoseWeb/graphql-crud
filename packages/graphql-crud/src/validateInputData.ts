import {
  getNullableType,
  GraphQLObjectType,
  GraphQLSchema,
} from 'graphql';

import {
  isEmpty,
  isPlainObject,
} from 'lodash';

import {
  isNonNullable,
} from './';

export interface ValidateInputDataProps {
  type: GraphQLObjectType;
  schema: GraphQLSchema;
  data: object;
}

// For every null value in the input data
// check that it can be nullable by check the type definition.
export const validateInputData = (props: ValidateInputDataProps) => {
  if (isEmpty(props.data)) {
    throw new Error('data input object is missing');
  }

  const fields = props.type.getFields();

  Object
    .keys(props.data)
    .forEach((key) => {
      const field = fields[key];
      const value = props.data[key];
      if (isPlainObject(value)) {
        validateInputData({
          schema: props.schema,
          data: value,
          type: getNullableType(field.type) as any,
        });
      } else {
        if (value === null && isNonNullable(field)) {
          throw new Error(`${props.type.name}.${field.name} must not be null`);
        }
      }
    });
};
