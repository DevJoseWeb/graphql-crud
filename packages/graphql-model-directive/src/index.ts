export * from './ModelDirective';
export * from './RelationDirective';
export * from './DefaultDirective';
export * from './UniqueDirective';
export * from './Storage';
export * from './generateFieldNames';

import { DefaultDirective } from './DefaultDirective';
import { ModelDirective } from './ModelDirective';
import { RelationDirective } from './RelationDirective';
import { UniqueDirective } from './UniqueDirective';

export default {
  model: ModelDirective,
  default: DefaultDirective,
  relation: RelationDirective,
  unique: UniqueDirective,
};
