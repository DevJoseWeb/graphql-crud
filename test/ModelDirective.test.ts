import { printSchema } from 'graphql';
import { makeExecutableSchema } from 'graphql-tools';
import { ModelDirective } from '../src';

const baseTypeDefs = `
  type Query {
    _: Boolean
  }

  type Mutation {
    _: Boolean
  }

  schema {
    query: Query
    mutation: Mutation
  }
`;

describe('ModelDirective', () => {
  const typeDefs = `
  type Foo @model {
    name: String
  }
  ${baseTypeDefs}
`;

  const schema = makeExecutableSchema({
    typeDefs,
    directives: {
      model: ModelDirective,
    } as any,
  });

  it('produces the expected schema', () => {
    expect(printSchema(schema)).toMatchSnapshot();
  });
});
