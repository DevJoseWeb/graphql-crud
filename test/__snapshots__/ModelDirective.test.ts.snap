// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelDirective produces the expected schema 1`] = `
"type Foo {
  name: String

  \\"\\"\\"Unique ID\\"\\"\\"
  id: ID
}

type Mutation {
  _: Boolean

  \\"\\"\\"Create a Foo\\"\\"\\"
  createFoo(data: Foo!): Foo

  \\"\\"\\"Update a Foo\\"\\"\\"
  updateFoo(data: Foo!): Foo

  \\"\\"\\"Update a Foo or create it if it doesn't exist\\"\\"\\"
  upsertFoo(data: Foo!): Foo

  \\"\\"\\"Remove a Foo\\"\\"\\"
  removeFoo: Foo
}

type Query {
  _: Boolean

  \\"\\"\\"Find one Foo\\"\\"\\"
  foo: Foo

  \\"\\"\\"Find multiple Foos\\"\\"\\"
  foos: Foo
}
"
`;
